File: src/main/java/com/github/zyypj/booter/updater/AutoUpdater.java
package com.github.zyypj.booter.updater;

import com.github.zyypj.booter.http.Request;

import java.io.*;
import java.net.URL;
import java.util.logging.Logger;

public class AutoUpdater {

    private final String currentVersion;
    private final String updateCheckUrl;
    private final String downloadUrl;
    private final File pluginFile;
    private final Logger logger;

    /**
     * @param currentVersion   A versão atual do plugin.
     * @param updateCheckUrl   A URL para verificar a versão mais recente.
     * @param downloadUrl      A URL para baixar o novo arquivo.
     * @param pluginFile       O arquivo atual do plugin (JAR).
     * @param logger           O logger para saída de logs.
     */
    public AutoUpdater(String currentVersion, String updateCheckUrl, String downloadUrl, File pluginFile, Logger logger) {
        this.currentVersion = currentVersion;
        this.updateCheckUrl = updateCheckUrl;
        this.downloadUrl = downloadUrl;
        this.pluginFile = pluginFile;
        this.logger = logger;
    }

    /**
     * Inicia o processo de verificação e atualização.
     */
    public void checkAndUpdate() {
        try {
            logger.info("Verificando atualizações...");

            String latestVersion = getLatestVersion();
            if (latestVersion == null) {
                logger.warning("Não foi possível verificar a versão mais recente.");
                return;
            }

            if (isUpToDate(latestVersion)) {
                logger.info("O plugin já está na versão mais recente: " + currentVersion);
                return;
            }

            logger.info("Nova versão encontrada: " + latestVersion + ". Iniciando atualização...");

            if (downloadUpdate()) {
                logger.info("Atualização concluída com sucesso. Reinicie o servidor para aplicar.");
            } else {
                logger.warning("Falha ao baixar a atualização.");
            }

        } catch (Exception e) {
            logger.severe("Erro durante o processo de atualização: " + e.getMessage());
            e.printStackTrace();
        }
    }

    /**
     * Verifica a versão mais recente no servidor.
     *
     * @return A versão mais recente, ou null se houver falha.
     */
    private String getLatestVersion() {
        Request<String> request = new Request<>(updateCheckUrl);
        if (request.execute()) {
            return request.getResponse().getRawData().trim();
        }
        return null;
    }

    /**
     * Verifica se a versão atual está atualizada.
     *
     * @param latestVersion A versão mais recente disponível.
     * @return true se estiver atualizado, false caso contrário.
     */
    private boolean isUpToDate(String latestVersion) {
        return currentVersion.equalsIgnoreCase(latestVersion);
    }

    /**
     * Baixa a nova versão do plugin.
     *
     * @return true se o download for bem-sucedido, false caso contrário.
     */
    private boolean downloadUpdate() {
        try (BufferedInputStream in = new BufferedInputStream(new URL(downloadUrl).openStream());
             FileOutputStream fileOutputStream = new FileOutputStream(new File(pluginFile.getParent(), "update-" + pluginFile.getName()))) {

            byte[] dataBuffer = new byte[1024];
            int bytesRead;

            while ((bytesRead = in.read(dataBuffer, 0, 1024)) != -1) {
                fileOutputStream.write(dataBuffer, 0, bytesRead);
            }

            logger.info("Atualização baixada com sucesso para: " + pluginFile.getParent());
            return true;
        } catch (IOException e) {
            logger.severe("Erro ao baixar a atualização: " + e.getMessage());
            e.printStackTrace();
            return false;
        }
    }
}

File: shared/src/main/java/me/zyypj/booter/shared/updater/AutoUpdater.kt
package me.zyypj.booter.shared.updater

import me.zyypj.booter.shared.http.client.HttpRequest
import java.io.BufferedInputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.net.URL
import java.util.logging.Logger

class AutoUpdater(
    private val currentVersion: String,   // Versão atual instalada
    private val updateCheckUrl: String,   // Endpoint que retorna a última versão disponível (texto simples)
    private val downloadUrl: String,      // URL do JAR mais recente
    private val pluginFile: File,         // Arquivo .jar atualmente em uso
    private val logger: Logger            // Logger fornecido pelo chamador
) {

    /**
     * Inicia a verificação e, se necessário, realiza o download da atualização.
     * Pode ser invocado de Java com:  new AutoUpdater(...).checkAndUpdate();
     */
    fun checkAndUpdate() {
        try {
            logger.info("Verificando atualizações…")
            val latestVersion = getLatestVersion()
            if (latestVersion == null) {
                logger.warning("Não foi possível verificar a versão mais recente.")
                return
            }

            if (isUpToDate(latestVersion)) {
                logger.info("O plugin já está na versão mais recente: $currentVersion")
                return
            }

            logger.info("Nova versão encontrada: $latestVersion. Iniciando atualização…")
            if (downloadUpdate()) {
                logger.info("Atualização concluída com sucesso. Reinicie o servidor para aplicar.")
            } else {
                logger.warning("Falha ao baixar a atualização.")
            }
        } catch (e: Exception) {
            logger.severe("Erro durante o processo de atualização: ${e.message}")
            e.printStackTrace()
        }
    }

    /**
     * Consulta o servidor e devolve a última versão em formato de string.
     * @return última versão ou *null* caso a requisição falhe.
     */
    private fun getLatestVersion(): String? {
        val request = HttpRequest<String>(updateCheckUrl)
        return if (request.execute()) request.response?.rawData?.trim() else null
    }

    /**
     * Compara a versão local com a versão disponível.
     * A comparação é *case‑insensitive* e sem análise semântica.
     */
    private fun isUpToDate(latestVersion: String): Boolean = currentVersion.equals(latestVersion, ignoreCase = true)

    /**
     * Faz o download do novo JAR para o mesmo diretório do plugin, com prefixo "update-".
     * @return *true* se o arquivo foi baixado sem erros.
     */
    private fun downloadUpdate(): Boolean {
        val targetFile = File(pluginFile.parentFile, "update-${pluginFile.name}")
        return try {
            BufferedInputStream(URL(downloadUrl).openStream()).use { input ->
                FileOutputStream(targetFile).use { output ->
                    val buffer = ByteArray(1024)
                    var bytesRead: Int
                    while (input.read(buffer).also { bytesRead = it } != -1) {
                        output.write(buffer, 0, bytesRead)
                    }
                }
            }
            logger.info("Atualização baixada com sucesso para: ${targetFile.absolutePath}")
            true
        } catch (e: IOException) {
            logger.severe("Erro ao baixar a atualização: ${e.message}")
            e.printStackTrace()
            false
        }
    }
}